modules/dataquery/php/queries.class.inc

```diff
@@ -0,0 +1,356 @@
+<?php
+namespace LORIS\dataquery;
+
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Message\ResponseInterface;
+use \LORIS\Data\Filters\AccessibleResourceFilter;
+
+/**
+ * Handles requests to queries under the /queries/* endpoint of
+ * the dataquery module.
+ *
+ * @license http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ */
+class Queries extends \NDB_Page
+{
+    public $skipTemplate = true;
+    /**
+     * Allow all users for testing
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $user   = $request->getAttribute('user');
+        $url    = $request->getURI()->getPath();
+        $pieces = [];
+        if (preg_match(
+            "/queries$/",
+            $url,
+            $pieces
+        ) === 1
+        ) {
+            switch ($request->getMethod()) {
+            case 'GET':
+                return new \LORIS\Http\Response\JSON\OK(
+                    [
+                        'queries' => iterator_to_array(
+                            $this->getUserAccessibleQueries($user),
+                            false, // do not preserve keys, stay an array
+                        ),
+                    ]
+                );
+            case 'POST':
+                $requestjson = (string) $request->getBody();
+                $requestdata = json_decode($requestjson, true);
+                if ($requestdata === false) {
+                    return new \LORIS\Http\Response\JSON\BadRequest(
+                        'Invalid JSON data'
+                    );
+                }
+                // FIXME: Validate query before storing it. Must contain fields.
+                $queryID = $this->_storeQuery($requestjson);
```

Comment:
This should not succeed if the fields or criteria are invalid.

Reply:
Should be fixed now